File: queues/C/src/queues.c
#include <stdlib.h>
#include <stdbool.h>
#include "queues.h"

void enqueue(Queue* queue, int element) {
    if (queue == NULL) {
        return;
    };

    Node* new_element = malloc(sizeof(Node));
    new_element->data = element;
    new_element->next = NULL;
    if (queue->size == 0) {
        queue->back = new_element;
        queue->front = new_element;
        queue->size++;
        return;
    };

    queue->back->next = new_element;
    queue->back = new_element;
    queue->size++;
    return;
};


bool dequeue(Queue* queue, int* dequeued_value) {
    if (queue == NULL || queue->size == 0) {
        return false;
    };

    *dequeued_value = queue->front->data;
    Node* next = queue->front->next;
    free(queue->front);
    queue->front = next;
    if (queue->size == 1) {
        queue->back = next;
    };
    queue->size--;
    return true;
};


bool peek(Queue* queue, int* peeked_value) {
    if (queue == NULL || queue->front == NULL) {
        return false;
    };

    *peeked_value = queue->front->data;
    return true;
};

bool isEmpty(Queue* queue) {
    if (queue == NULL) {
        return true;
    };

    return (queue->size == 0);
};


int queueSize(Queue* queue) {
    if (queue == NULL) {
        return 0;
    };

    return queue->size;
};

File: queues/Py/queues.py
class Node:
    def __init__(self, data: int):
        self.data = data
        self.next: Node | None = None


class Queue:
    def __init__(self):
        self.front: Node | None = None
        self.back: Node | None = None
        self._size = 0
    

    def enqueue(self, element: int) -> None:
        new_node = Node(element)
        if self._size == 0:
            self.front = new_node
            self.back = new_node
            self._size+=1
            return
        assert self.back is not None
        self.back.next = new_node
        self.back = new_node
        self._size+=1
        return
    
    def dequeue(self) -> int | None:
        if self._size == 0:
            return None
        assert self.front is not None
        dequeued_value = self.front.data
        next = self.front.next
        self.front = next
        self._size-=1
        return dequeued_value
    
    def peek(self) -> int | None:
        if self._size == 0:
            return None
        assert self.front is not None
        return self.front.data
    
    def isEmpty(self) -> bool:
        return (self._size == 0)
    
    def queueSize(self) -> int:
        return self._size
    

